### PR TITLE
feat: darker icons for light colorschemes

### LINKED ISSUE #193: [Proposal] Set different icon color for dark/light background respectively
Is it possible to use a different set of colors for light backgrounds? Currently most icons looks great under dark background but some of them are not displayed clearly under light background due to their color:

image
image

### PR DESCRIPTION
**Pull request by mg979**
fixes #193

### CODE REVIEW

**Comment by mg979** [comment]
I'll add some tolerance in the color change, because actually only very light colors need to be darkened.

**Comment by mg979** [comment]
I don't invert them anymore, they are just darkened, I think it's pretty decent this way.

File: lua/nvim-web-devicons.lua [hunk 1]
```diff
@@ -1539,8 +1539,25 @@ local icons_by_file_extension = {
   },
 }
 
+local icons
+
+local function refresh_icons()
```
Comment (alex-courtis) [inline lua/nvim-web-devicons.lua:1544]: We can make this API

Please add doc to readme

File: lua/nvim-web-devicons.lua [hunk 2]
```diff
@@ -1539,8 +1539,25 @@ local icons_by_file_extension = {
...
+  end
+  icons = vim.tbl_extend("keep", {}, by_filename, by_file_extension)
+end
+
 -- When adding new icons, remember to add an entry to the `filetypes` table, if applicable.
-local icons = vim.tbl_extend("keep", {}, icons_by_filename, icons_by_file_extension)
+refresh_icons()
+
+-- Change icon set on background change
+vim.api.nvim_create_autocmd("OptionSet", { pattern = "background", callback = refresh_icons })
```
Comment (alex-courtis) [inline lua/nvim-web-devicons.lua:1560]: Please move this new code down to the bottom. It's "init code" so fits nicely just before the final return.

File: scripts/darken_colors.lua [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    r = bit.tohex(r / 3):sub(-2)
+    g = bit.tohex(g / 3):sub(-2)
+    b = bit.tohex(b / 3):sub(-2)
+  else ------------------------------------- lightest octile
+    r = bit.tohex(r / 5):sub(-2)
+    g = bit.tohex(g / 5):sub(-2)
+    b = bit.tohex(b / 5):sub(-2)
+  end
+  return string.format("#%s%s%s", r, g, b)
+end
```
Comment (alex-courtis) [inline scripts/darken_colors.lua:53]: Could we perhaps go a bit darker? This looks good (base16-default-light with #ffffff background) however some are still a little indistinct:

File: scripts/cterm-colors-generator.lua [hunk 4]
```diff
@@ -10,12 +10,12 @@
 
 local fn = vim.fn
 
-vim.cmd("drop lua/nvim-web-devicons.lua")
-
-if not fn.bufname():find("nvim%-web%-devicons%.lua$") then
-  error("The file could not be found!")
+if fn.filereadable("lua/nvim-web-devicons.lua") == 0 then
+  error("Your working directory isn't set to correctly.")
```
Comment (alex-courtis) [inline scripts/cterm-colors-generator.lua:14]: /nit

error("lua/nvim-web-devicons.lua not found")

**Review by alex-courtis — CHANGES_REQUESTED** [review]
This works beautifully.

Please:

- [ ] add API
- [ ] refactor new functions in nvim-web-devicons.lua

We now must add automation / documentation so that future contributions run this. We should also always run `cterm-colors-generator.lua`

Please:

- [ ] Create a shell script that can run both generators. Make a start, I can assist with this.

We can document this all at the end, when everything is complete and ready. CONTRIBUTING.md should be appropriate, with optional CI.

**Comment by alex-courtis** [comment]
This works beautifully. The resulting script is well formatted too.

File: scripts/darken_colors.lua (hunk 3 above)
Comment (mg979) [reply scripts/darken_colors.lua:53]: You mean the folder icon or which ones?

**Comment by mg979** [comment]
I removed the old script, now there's only one, the whole thing can be run from the shell:

    make colors

The icons should be a bit darker too, but not too much. The `folder` icon used by NvimTree seems to be their own, it's not in nvim-devicons, so it cannot be made darker with this script.

**Comment by mg979** [comment]
With this setup there's only `make colors` to run.

CONTRIBUTING.md should mention that:

- after adding new icons `make colors` should be run
- lifepillar/colortemplate should be installed and in runtimepath (command would fail without it)

Can you create CONTRIBUTING.md? I don't have a template for that, nor I know exactly what would go in there.

File: scripts/darken_colors.lua (hunk 3 above)
Comment (alex-courtis) [reply scripts/darken_colors.lua:63]: I was thinking about config.mk. The folder ones can be ignored as they are from nvim-tree.

Don't worry for now; I might run a test with all the icons and see how they all look.

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,4 @@
+colors:
+	nvim --headless -c 'source scripts/darken_colors.lua' -c 'qall'
+
+all: colors
```
Comment (alex-courtis) [inline Makefile:4]: Very nice.

File: scripts/darken_colors.lua [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  error("Neovim must be LuaJIT-enabled to source this script")
+end
+
+if fn.filereadable("lua/nvim-web-devicons.lua") == 0 then
+  error("lua/nvim-web-devicons.lua not found")
+end
+
+if not vim.o.runtimepath:find("vim%-colortemplate") then
+  error("lifepillar/vim-colortemplate must be present in the runtimepath.")
+end
```
Comment (alex-courtis) [inline scripts/darken_colors.lua:22]: Not happy when vim-colortemplate is at `~/.local/share/nvim/site/pack/packer/start/vim-colortemplate`.

How about:
```
local rc, err = pcall(vim.fn["colortemplate#colorspace#approx"], "#000000")
if not rc then
  error(err .. "\nPlease ensure lifepillar/vim-colortemplate present in the runtimepath.")
end
```

File: scripts/darken_colors.lua [hunk 7]
```diff
@@ -0,0 +1,164 @@
+-- Plugin lifepillar/vim-colortemplate must be available on &runtimepath
```
Comment (alex-courtis) [inline scripts/darken_colors.lua:1]: Let's rename this to `generate_colours.lua`.

I like having everything in one place.

**Comment by alex-courtis** [comment]
> Can you create CONTRIBUTING.md? I don't have a template for that, nor I know exactly what would go in there.

Done. Please review.

I'll deal with code style later as it will create noise in this PR.

CI may be impossible. We can address that if it ever becomes a problem.

File: lua/nvim-web-devicons.lua (hunk 1 above)
Comment (alex-courtis) [reply lua/nvim-web-devicons.lua:1546]: We don't really need this. Ignore me.

**Review by alex-courtis — CHANGES_REQUESTED** [review]

**Comment by alex-courtis** [comment]
Very happy with the outcome here!

**Review by alex-courtis — APPROVED** [review]
Incredible!

This is a fantastic contribution and greatly improves functionality and developability.

Many, many thanks.

**Pull request merged**